### DISCUSSION TITLE
Now support web apps

### DISCUSSION THREAD

**Discussion by NanoMichael** [NanoMichael/MicroTeX#93] [Announcements]
Now we support web apps, check [this demo](https://nanomichael.github.io/MicroTeX/) out.

And I'm going to release version 1.0.0, but before that, we have a lot of non-trivial work to do.

And most important, I'm planning to rename the project to `TinyTeX`, as its main goal is: fast and tiny. But I'm wondering if is it appropriate. Can you guys give me some advice? @sp1ritCS @jeremysanders @blackhole89 @PikachuHy 👀

**Comment by PikachuHy** [comment]
good job.
I will review code.

**Comment by PikachuHy** [comment]
- Do we need to rename all `LaTeX` to `TinyTex`?

https://github.com/NanoMichael/cLaTeXMath/blob/dc32540664276a00805b21a0fb5cd6418bacdb54/CMakeLists.txt#L2

https://github.com/NanoMichael/cLaTeXMath/blob/ae70d932fb78e45e2c8a8e029ee95fe52820ee77/platform/qt/CMakeLists.txt#L3



- Confusing although compiled.

https://github.com/NanoMichael/cLaTeXMath/blob/ae70d932fb78e45e2c8a8e029ee95fe52820ee77/lib/otf/clm.cpp#L24
What does this code mean? convert `(*(p + i))` to `T` or convert `(*(p + i)) << ((shift - i) * 8)` to `T`?

comiled with warning
```
C:\code\cLaTeXMath\lib\otf\clm.cpp(24): warning C4805: '|=': unsafe mix of type 'T' and type 'int' in operation
        with
        [
            T=bool
        ]
C:\code\cLaTeXMath\lib\otf\clm.cpp(89): note: see reference to function template instantiation 'T tex::BinaryReader::read<bool>(void)' being compiled
        with
        [
            T=bool
        ]
C:\code\cLaTeXMath\lib\otf\clm.cpp(24): warning C4805: '|': unsafe mix of type 'bool' and type 'int' in operation
```

- Add conan support

public our library to conan-center: https://conan.io/center/

[Conan](https://conan.io/) is an open source, decentralized and multi-platform package manager to create and share all native binaries.


- suggestion: Add prebuilt clm files to v1.0.0 release package. (e.g. XITSMath-Regular.otf)
- suggestion: Use a separate GitHub organization. (e.g. tinytex)

**Reply by NanoMichael** [reply]
> What does this code mean? convert `(*(p + i))` to `T` or convert `(*(p + i)) << ((shift - i) * 8)` to `T`?

It is to convert a byte sequence to a certain primitive type like to convert a `float` from a byte sequence, this warning will happen when we attempt to convert a `bool` from that byte sequence, but I believe this can be ignored 🤔

**Reply by NanoMichael** [reply]
> Do we need to rename all LaTeX to TinyTex?

Yes. it is overstated that we name it `LaTeX` 😄 , this project is not full `LaTeX` supported actually...

**Reply by PikachuHy** [reply]
can we add a static_cast to eliminate the warning? 😂

**Reply by PikachuHy** [reply]
I will change all LaTeX to TinyTeX in next PR.

**Reply by PikachuHy** [reply]
> change LaTeX to TinyTeX

see https://github.com/NanoMichael/cLaTeXMath/pull/98.

a not pure PR. (⊙﹏⊙)

**Reply by sp1ritCS** [reply]
> * Add conan support
> public our library to conan-center: https://conan.io/center/
> Conan is an open source, decentralized and multi-platform package manager to create and share all native binaries.

never heard of it. but given that it isn't in any official repository on almost all linuxes, it doesn't seem like something to prefer to native packages. Especially since conan just provides a "linux" build, instead of one for each distro, there will be symbol incompatibilities of dependencies.

---

> suggestion: Add prebuilt clm files to v1.0.0 release package. (e.g. XITSMath-Regular.otf)

please don't, @LyesSaadi and me would like to avoid any licensing issues that would come with that. Instead we'll pull the otf files from their respective packages (i.e. texlive-xits-fonts) and build the clm file during buildtime.

---

> suggestion: Use a separate GitHub organization. (e.g. tinytex)

why? it still seems like @NanoMichael s project, so why would you need to separate it into a separate org?

**Reply by LyesSaadi** [reply]
> please don't, @LyesSaadi and me would like to avoid any licensing issues that would come with that. Instead we'll pull the otf files from their respective packages (i.e. texlive-xits-fonts) and build the clm file during buildtime.

oh no, halp, no licensing issues. please.

**Reply by PikachuHy** [reply]
> conan just provides a "linux" build

conan is written by python.
It is just a package manager like vcpkg, which we used in ci.

from conan doc
> Conan is a dependency and package manager for C and C++ languages. It is free and open-source, works in all platforms ( Windows, Linux, OSX, FreeBSD, Solaris, etc.), and can be used to develop for all targets including embedded, mobile (iOS, Android), and bare metal. It also integrates with all build systems like CMake, Visual Studio (MSBuild), Makefiles, SCons, etc., including proprietary ones.

maybe we can also add brew support (for mac user).

**Reply by sp1ritCS** [reply]
> conan is written by python.
> It is just a package manager like vcpkg, which we used in ci.

Yeah, but that's not what I ment. Besides "ALT Linux", there is no Linux distro which officially offers conan (https://pkgs.org/search/?q=conan). That leads me to asume that it's not that popular (even more so the fact, as they do not even ship a gtkmm package there).

Additionally, the problem is, that they do not offer builds for various distros, but only one "Linux". But obv. different Linuxes may ship with different versions of libraries and thus there might be symbols that will be missing on some distros. 

> maybe we can also add brew support (for mac user).

see: https://github.com/sp1ritCS/homebrew-tap/blob/master/Formula/clatexmath.rb

**Reply by PikachuHy** [reply]
> there is no Linux distro which officially offers conan

yes

conan is installed by `pip install conan`

just provide a option to our user.
native package manager(apt or yum) or 3rd package(vcpkg, conan, etc) is ok.

**Reply by sp1ritCS** [reply]
> conan is installed by `pip install conan`

yes, but that is obv. not officially offered. instead if pulls the conan source of pypi which is not validated by any of the distros.

if it was popular, there'd be some kind of (python38-)conan package in the official repos, like pip itself is (python38-pip). 

I'm just considering if the effort of publishing it as a package on a platform that nobody uses 😦, but if I want to I don't think anybody will stop you :).

**Comment by PikachuHy** [comment]
I met a error.

see https://github.com/NanoMichael/cLaTeXMath/issues/97

**Reply by NanoMichael** [reply]
Seems that the version of the `fontforge` is too low? By the way, now we can find the prebuilt `clm` resources from [here](https://github.com/NanoMichael/cLaTeXMath/tree/openmath/res).

**Reply by PikachuHy** [reply]
only otf files here

**Reply by NanoMichael** [reply]
I'll check it soon

**Reply by NanoMichael** [reply]
> only otf files here

fixed by 2bf07ba

**Comment by LyesSaadi** [comment]
:O, the web app demo looks so cool (and works quite well)! I guess it's through WASM? (just checked, it seems so)

The TinyTex rename doesn't have much impact on packaging for Fedora since nothing was pushed yet (and, anyway, renames are not hard to handle). Also, the new name seems to be more discoverable (I'd image someone would just type `tiny latex library` or `tiny tex library` to find it).

**Comment by NanoMichael** [comment]
Well, I've found a project (see [here](https://github.com/yihui/tinytex)) already named `TinyTeX` :joy:... there are some choices:

- FastTeX, sounds a little bit silly
- MathTeX or TeXMath, not bad
- ~~TinyMath, already have one, see [here](https://github.com/elastic/tinymath)~~

Any suggestions?

**Reply by sp1ritCS** [reply]
I don't like either of them that much, what about NativeTeX or Quick(La)?TeX ?

**Reply by jeremysanders** [reply]
EmbedTeX, MicroTeX, UsefulTeX?

**Reply by NanoMichael** [reply]
You guys are genius! QuickTeX and MicroTeX sound excellent, and I prefer `MicroTeX`, ~~or `MiniTeX` (well, there already have one), or let me be selfish, what about `NanoTeX`? :joy: (unfortunately, there also already have one)~~

**Reply by NanoMichael** [reply]
Well, there is already have a project also named `MicroTeX`... What about XTeX, `X` means cross-platform :thinking:

**Reply by sp1ritCS** [reply]
X usually means the X Window System/XFree86/Xorg stack on Unix at least

that would be confusing I think

**Reply by NanoMichael** [reply]
That makes sense :thinking: After some searches, I've found there is more than one project already named `QuickTeX`. And I've found the project [microtex](https://github.com/gitpetrol/microtex) is nothing about `TeX`, I think `MicroTeX` is good.

**Reply by PikachuHy** [reply]
`NanoTeX` is the best.

**Reply by NanoMichael** [reply]
> `NanoTeX` is the best.

I like it too, but there's also already have one, see [here](https://github.com/corwin-of-amber/nanotex).

**Reply by sp1ritCS** [reply]
yeah, {pico,femto}tex seems to small and the only microtex has nothing to do with LaTeX. I don't think there is risk of confusing, esp. if we write it like \mu\TeX.

**Reply by NanoMichael** [reply]
WOW! `\mu\TeX` is amazing :tada:! So, what about we renaming the `namespace` to `microtex` (`mutex` is confusing with `std::mutex`) and `TinyTeX` to `MuTeX` (or just `MicroTeX`)?

**Reply by sp1ritCS** [reply]
TinyTeX currently is only in strings, isn't it? Maybe then we could also use the Unicode µ (so µTeX) in those places, or is that to risky?

**Reply by NanoMichael** [reply]
Do you mean we name the project `µTeX`, but keep the `tinytex` namespace? I believe it is a GENIUS idea, we can explain to the users that 'µ' stands for tiny.

**Reply by sp1ritCS** [reply]
µ stands for micro not tiny?

**Reply by NanoMichael** [reply]
:thinking: yeah, that's a problem, seems the renaming is inevitable.

**Reply by NanoMichael** [reply]
Or rename the namespace to `meutex`, to distinguish from `mutex` :joy:

**Reply by sp1ritCS** [reply]
No I'd say keem microtex as namespace, \mu\TeX for in-renderer name and µTeX for the string name. If you do not want non-ascii chars in strings, I'd say go with MicroTeX for the string name.

**Reply by LyesSaadi** [reply]
Uhhh, I'm quite confident that unicode characters will lead to unforeseen problems in the future :sweat_smile:...

**Reply by NanoMichael** [reply]
yeah, it is too hard for those non-Greek users to input the 'µ', but 'µTeX' as the logo is great. Do we consider using `meutex`? :thinking:

**Reply by sp1ritCS** [reply]
I have no idea why you'd call it **meu**tex. I'd say name it microtex, given that µ is the SI prefix for micro (`10^{-6}`).

Also I think all ANSI keyboards have the ability to input µ with AltGr+M (I think), but yeah might be hard for Americans :)

**Reply by NanoMichael** [reply]
> I have no idea why you'd call it meutex.

It is a homonym for 'mu' :joy: I think microtex is good. Please see [here](https://nanomichael.github.io/cLaTeXMath/?tex=%22\text{Hello%20from%20\mu\kern-.1em\TeX}%22), I got your point.

**Reply by sp1ritCS** [reply]
> It is a homonym for 'mu' :joy:

huh, then maybe pronunciation varies in different languages/cultures. In Germany we say mü (IPA [myː]) not [mɔʏ̯] or [meuː]

**Reply by NanoMichael** [reply]
> huh, then maybe pronunciation varies in different languages/cultures. In Germany we say mü (IPA [myː]) not [mɔʏ̯] or [meuː]

Oh, I didn't realize it at all. 

So, the answer becomes clear, in conclusion, use 'microtex' as namespace, and \mu\TeX for the in-renderer name.  I'll make a PR for renaming.

**Comment by PikachuHy** [comment]
> And I'm going to release version 1.0.0

Some problems I found when studying or reviewing this project.

**Reply by PikachuHy** [reply]
suggestion: Add unit test

**Comment by PikachuHy** [comment]
- [x] typo

https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/atom/atom_char.h#L20

**Comment by PikachuHy** [comment]
- [ ] Logical expression is always true

https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/core/parser.cpp#L719

**Comment by PikachuHy** [comment]
- [x] unnecessary `L`

After using `std::string`, `L` is redundant.

https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/core/parser.cpp#L266

https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/core/parser.cpp#L279

**Comment by PikachuHy** [comment]
- [x] refactor: use `isAlpha` in utils.h

https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/core/parser.cpp#L270

**Comment by PikachuHy** [comment]
- [x] typo 

everywhere???


https://github.com/NanoMichael/MicroTeX/blob/9cae2405fd2ab9bbfdfaa9ecefde511d32e80497/lib/atom/atom_row.h#L140
